### PR TITLE
chore(mlx-lm): fix tie_word_embeddings for qwen2

### PR DESCRIPTION
**Pull request by mzbac**
fix: https://github.com/ml-explore/mlx-examples/issues/564

### CODE REVIEW

**Comment by awni** [comment]
There is one thing I'm unsure about here and that is what happens when we quantize the model?

If `tie_word_embeddings = True` and we add a linear output layer, then that output layer will be quantized. I think it should work in the sense that the quantized output weights will be saved and then we will use them directly. But maybe that is not the desired behavior?

**Comment by mzbac** [comment]
> There is one thing I'm unsure about here and that is what happens when we quantize the model?
> 
> If `tie_word_embeddings = True` and we add a linear output layer, then that output layer will be quantized. I think it should work in the sense that the quantized output weights will be saved and then we will use them directly. But maybe that is not the desired behavior?

Would you please elaborate on your concern a bit? It makes sense to me that once we save the quant weights which include the cloned lm_head, then when we load the quant model with saved the cloned lm_head, the model should work as expected.

I think maybe we should default tie_word_embeddings to true in the model arguments, since the transformers default tie_word_embeddings to be true. just in case some model configurations are missing the default tie_word_embeddings in the config file.

**Comment by awni** [comment]
> Would you please elaborate on your concern a bit? It makes sense to me that once we save the quant weights which include the cloned lm_head, then when we load the quant model with saved the cloned lm_head, the model should work as expected.

I think it's probably minor. But, what I was thinking is maybe it's unexpected if you have `tie_word_embeddings=True` that you have a quantized LM head which is separate from the embeddings. My expectation if that flag is true is that the two layers use the exact same weights (as in the same memory blob).

File: llms/tests/test_models.py [hunk 1]
```diff
@@ -129,6 +129,47 @@ def test_qwen2(self):
             model, args.model_type, args.vocab_size, args.num_hidden_layers
         )
 
+    def test_qwen2_tie_word_embeddings_without_lm_head_weight(self):
```
Comment (awni) [inline llms/tests/test_models.py:132]: Thanks for adding some tests ❤️

**Comment by mzbac** [comment]
> > Would you please elaborate on your concern a bit? It makes sense to me that once we save the quant weights which include the cloned lm_head, then when we load the quant model with saved the cloned lm_head, the model should work as expected.
> 
> I think it's probably minor. But, what I was thinking is maybe it's unexpected if you have `tie_word_embeddings=True` that you have a quantized LM head which is separate from the embeddings. My expectation if that flag is true is that the two layers use the exact same weights (as in the same memory blob).

My thought is that some models have already cloned the token embedding to lm_head in their released weights (e.g. Qwen2 0.5B for instances), and our quantization seems to work fine with that case. I am not sure if we still need to keep those two layers the same by introducing some checks in the quantization part.

**Review by awni — APPROVED** [review]
Thanks for the fix! 🚀

**Pull request merged**